bootloader/grub.go

```diff
@@ -440,6 +455,39 @@ func (g *grub) CandidateCommandLine(pieces CommandLineComponents) (string, error
 	return g.commandLineForEdition(edition, pieces)
 }
 
+// DefaultCommandLine returns the default kernel command-line used by
+// the bootloader excluding the recovery mode and system parameters.
+func (g *grub) DefaultCommandLine(candidate bool) (string, error) {
+	var edition uint
+
+	// if "candidate", we look for the managed boot assets
+	// (current snapd) rather than the ones currently installed on
+	// the boot/seed disk. This is needed to know the default
+	// command line before candidate boot assets are installed
+	if candidate {
+		assetName := "grub.cfg"
+		if g.recovery {
+			assetName = "grub-recovery.cfg"
+		}
+
+		var err error
+		edition, err = editionFromInternalConfigAsset(assetName)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		currentBootConfig := filepath.Join(g.dir(), "grub.cfg")
```

Comment:
why do we not account for `g.recovery` here?

Reply:
There are 4 different `grub.cfg`.
 * `grub.cfg` installed in the seed partition
 * `grub.cfg` installed in the boot partition
 * `grub.cfg` from snapd to be installed to the boot partition
 * `grub-recovery.cfg` from snapd to be installed on the seed partition as `grub.cfg`

Also note that the two latter from snapd (the candidates) are not file on the system. They are files in the source of snapd at `bootloader/assets/data/`. But they are injected in go code so that they are just strings in the binary. At no point there is a file on the system called `grub-recovery.cfg`. It is just an identifier.

Reply:
Ah I see, thanks for explaining